x-pack/plugins/cloud_security_posture/public/pages/rules/rules_context.tsx

```diff
@@ -0,0 +1,365 @@
+        (e: CspBenchmarkRulesWithStates) => ({
+          benchmark_id: e?.metadata.benchmark.id,
+          benchmark_version: e?.metadata.benchmark.version,
+          rule_number: e?.metadata.benchmark.rule_number!,
+          rule_id: e?.metadata.id,
+        })
+      );
+      // Only do the API Call IF there are no undefined value for rule number in the selected rules
+      if (!bulkSelectedRules.some((rule) => rule.rule_number === undefined)) {
+        uiMetricService.trackUiMetric(METRIC_TYPE.COUNT, CHANGE_MULTIPLE_RULE_STATE);
+        mutateRuleState({
+          newState: state,
+          ruleIds: bulkSelectedRules,
+        });
+      }
+      setSelectedRules([]);
+    },
+    [selectedRules, mutateRuleState]
+  );
+
+  const setSelectAllRules = useCallback(() => {
+    setSelectedRules(filteredRulesWithStates);
+  }, [filteredRulesWithStates]);
+
+  const rulesFlyoutData: CspBenchmarkRulesWithStates = useMemo(() => {
+    const arrayRulesStates: RuleStateAttributes[] = Object.values(rulesStates.data || {});
+    return {
+      ...{
+        state:
+          arrayRulesStates.find((filteredRuleState) => filteredRuleState.rule_id === params.ruleId)
+            ?.muted === true
+            ? 'muted'
+            : 'unmuted',
+      },
+      ...{
+        metadata: allRules?.data?.items.find((rule) => rule.metadata.id === params.ruleId)
+          ?.metadata!,
+      },
+    };
+  }, [rulesStates, params.ruleId, allRules]);
+
+  // This useEffect is in charge of auto paginating to the correct page of a rule from the url params
+  useEffect(() => {
+    const getPageByRuleId = () => {
+      if (params.ruleId && allRules?.data?.items) {
+        const ruleIndex = allRules?.data?.items.findIndex(
+          (rule) => rule.metadata.id === params.ruleId
+        );
+
+        if (ruleIndex !== -1) {
+          // Calculate the page based on the rule index and page size
+          const rulePage = Math.floor(ruleIndex / pageSize);
+          return rulePage;
+        }
+      }
+      return 0;
+    };
+
+    setPage(getPageByRuleId());
+  }, [allRules?.data?.items, params.ruleId, pageSize]);
+
+  const contextValue = useMemo<RulesContextValue>(
+    () => ({
+      section,
+      setSection,
+      page,
+      pageSize,
+      setPageSize,
+      sortField,
+      setSortField,
+      sortOrder,
+      setSortOrder,
+      ruleNumber,
+      setRuleNumber,
+      search,
+      setSearch,
+      // rulesPageData,
+      loading,
+      error: error ? extractErrorMessage(error) : undefined,
+      total,
+      rules,
+      rulesShown,
+      setPage,
+      sectionList,
+      ruleNumberSelectOptions,
+      sectionSelectOptions,
+      selectedRules,
+      setSelectedRules,
+      setSelectAllRules,
+      setEnabledDisabledItemsFilter,
+      toggleRuleState,
+      toggleSelectedRulesStates,
+      enabledDisabledItemsFilter,
+      mutedRulesCount,
+      rulesFlyoutData,
+    }),
```

Comment:
You are right,  I am not sure either if useMemo even makes a difference at the moment.   I added this at first to follow the pattern I usually use, but since the Context values are all connected to a change in the fetched data it may not make sense.